---
title: "Whole_organoid_yap"
author: "Francesco Valzano"
date: "8/10/2024"
output: html_document
---
```{r libraries, include=FALSE}
library(Seurat)
library(SeuratDisk)
library(ggplot2)
library(harmony)
library(scutils2)
library(scater)
library(readr)
library(stringr)
library(reticulate)
library(SeuratWrappers)
library(DoubletFinder)

```
```{r setup}
#Load the data
base_directory = "/hpc/pmc_kool/fvalzano/Rstudio_Test1/Ependymoma_organoids/Whole/" 
filenames = list.files(paste0(base_directory, "seurat_objs/Post_preprocessing/"))
qc_seurat_objects = list()
for (i in filenames){
    qc_seurat_objects[[i]] = readRDS(paste0(base_directory, "seurat_objs/Post_preprocessing/", i))
}
names(qc_seurat_objects) = str_remove(filenames, ".rds")
```
```{r subset YAP organoids}
YAP = qc_seurat_objects[c("LX360", "LX362", "LX687_YAP", "LX687_Ctrl")]
```

```{r Doublet Identification}
YAP$LX687_YAP$genotype = NULL
YAP$LX687_Ctrl$genotype = NULL
for (i in names(YAP)) {
  # print the sample we are on
    print(paste0("Sample ",i))
  
  # Pre-process seurat object 
    YAP[[i]] = SCTransform(YAP[[i]], vars.to.regress = c("percent.mt", "percent.ribo"), vst.flavor = "v2", assay = "RNA")
    YAP[[i]]= CellCycleScoring(YAP[[i]], s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, assay = 'SCT')
    YAP[[i]] = SCTransform(YAP[[i]], vars.to.regress = c("percent.mt", "percent.ribo", 'S.Score', 'G2M.Score'), vst.flavor = "v2", assay = "RNA")
    hvg = VariableFeatures(YAP[[i]])
    ribo.genes <- grep(pattern = "^RP[SL]", x = hvg, value = TRUE)
    subtract<-which(hvg %in% ribo.genes)
    hvg<-hvg[-subtract]
    YAP[[i]] <- RunPCA(YAP[[i]], nfeatures.print = 10, features=hvg)
    
  # Find significant PCs
    stdv <- YAP[[i]][["pca"]]@stdev
    sum.stdv <- sum(YAP[[i]][["pca"]]@stdev)
    percent.stdv <- (stdv / sum.stdv) * 100
    cumulative <- cumsum(percent.stdv)
    co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
    co2 <- sort(which((percent.stdv[1:length(percent.stdv) - 1] - 
                       percent.stdv[2:length(percent.stdv)]) > 0.1), 
              decreasing = T)[1] + 1
    min.pc <- min(co1, co2)
    min.pc
  
  # finish pre-processing
    YAP[[i]] <- RunUMAP(YAP[[i]], dims = 1:min.pc)
    YAP[[i]] <- FindNeighbors(object = YAP[[i]], dims = 1:min.pc)              
    YAP[[i]] <- FindClusters(object = YAP[[i]], resolution = 0.1)
  
  # pK identification (no ground-truth)
    sweep.list <- paramSweep(YAP[[i]], PCs = 1:min.pc, sct=T)
    sweep.stats <- summarizeSweep(sweep.list)
    bcmvn <- find.pK(sweep.stats)
  
  # Optimal pK is the max of the bomodality coefficent (BCmvn) distribution
    bcmvn.max <- bcmvn[which.max(bcmvn$BCmetric),]
    optimal.pk <- bcmvn.max$pK
    optimal.pk <- as.numeric(levels(optimal.pk))[optimal.pk]
  
  ## Homotypic doublet proportion estimate
    annotations <- YAP[[i]]@meta.data$seurat_clusters
    homotypic.prop <- modelHomotypic(annotations) 
    nExp.poi <- round(optimal.pk * nrow(YAP[[i]]@meta.data)) ## Assuming 7.5% doublet formation rate - tailor for your dataset
    nExp.poi.adj <- round(nExp.poi * (1 - homotypic.prop))
  
  # run DoubletFinder
    YAP[[i]] <- doubletFinder(seu = YAP[[i]], 
                                   PCs = 1:min.pc, 
                                   pK = optimal.pk,
                                   nExp = nExp.poi.adj,
                                   sct=T)
    metadata <- YAP[[i]]@meta.data
    names(metadata)[16] <- "doublet_finder"    
    YAP[[i]]@meta.data <- metadata 
    df = as.data.frame(YAP[[i]]$doublet_finder)
    df$barcodes = colnames(YAP[[i]])
    colnames(df) = c("droplet_content", "barcodes")
    write_csv2(df, paste0("/hpc/pmc_kool/fvalzano/Rstudio_Test1/Ependymoma_organoids/Whole/DoubletFinder/", "DoubletFinder_",i,".csv"))
}
```

```{r Data Merging -Important in Seurat v5- SCT Normalization}
#In Seurat v5 first merge the objects and then integrate them
YAP_first = YAP[[1]]
YAP[[1]] = NULL
YAP =  merge(x = YAP_first, y= c(YAP), merge.data = TRUE, project = "Ep_whole_organoids") 
#Perform SCT normalization and then calculate cellcycle scoring, after that, rerun SCT regressing out the cell cycle Phase
scrna_YAP = SCTransform(YAP, vars.to.regress = c("percent.mt", "percent.ribo"), vst.flavor = "v2", assay = "RNA")
scrna_YAP = CellCycleScoring(scrna_YAP, s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, assay = 'SCT')
scrna_YAP = SCTransform(scrna_YAP, vars.to.regress = c("percent.mt", "percent.ribo", 'S.Score', 'G2M.Score'), vst.flavor = "v2", assay = "RNA")
```

```{r Dimensionality reduction}
#Remove potential ribosomial genes from high variable features used for PCA
hvg = VariableFeatures(scrna_YAP)
ribo.genes <- grep(pattern = "^RP[SL]", x = hvg, value = TRUE)
subtract<-which(hvg %in% ribo.genes)
hvg<-hvg[-subtract]
#Perform dimensionality reduction on unintegrated object
scrna_YAP<-RunPCA (scrna_YAP, verbose = T, features = hvg)
#Remove Ribosomial genes from variable features used for pca
scrna_YAP <- FindNeighbors(object = scrna_YAP, reduction = "pca", dims = 1:30)
scrna_YAP = RunUMAP(scrna_YAP, reduction = "pca", dims = 1:30, reduction.name = "umap")
i = seq(0.2, 1, by = 0.2)
scrna_YAP <- FindClusters(scrna_YAP, resolution = i)
```

```{r Harmony Integration}
scrna_YAP <- IntegrateLayers(object = scrna_YAP, 
                             method = HarmonyIntegration,
                             orig.reduction = "pca", 
                             normalization.method = "SCT", 
                             new.reduction = "harmony")
scrna_YAP <- FindNeighbors(scrna_YAP, reduction = "harmony", dims = 1:30)
scrna_YAP <- FindClusters(scrna_YAP, resolution = i, cluster.name=paste0("Harmony_SCT_",i))
scrna_YAP <- RunUMAP(scrna_YAP, reduction = "harmony", dims = 1:30, reduction.name = "umap_harmony")
```
```{r CCA Integration}
scrna_YAP <- IntegrateLayers(object = scrna_YAP, 
                             method = CCAIntegration,
                             orig.reduction = "pca", 
                             normalization.method = "SCT", 
                             new.reduction = "cca")
scrna_YAP <- FindNeighbors(scrna_YAP, reduction = "cca", dims = 1:30)
scrna_YAP <- FindClusters(scrna_YAP, resolution = i, cluster.name=paste0("CCA_SCT_",i))
scrna_YAP <- RunUMAP(scrna_YAP, reduction = "cca", dims = 1:30, reduction.name = "umap_cca")
scrna_YAP = JoinLayers(scrna_YAP, assay = "RNA")
```

```{r Exclude doublets}
write_rds(scrna_YAP, "/hpc/pmc_kool/fvalzano/Rstudio_Test1/Ependymoma_organoids/Whole/seurat_objs/Post_integration/scrna_YAP.rds")
Idents(scrna_YAP) = "doublet_finder"
scrna_YAP_noDoublets = subset(scrna_YAP, idents = "Singlet")
write_rds(scrna_YAP_noDoublets, "/hpc/pmc_kool/fvalzano/Rstudio_Test1/Ependymoma_organoids/Whole/seurat_objs/Post_integration/scrna_YAP_noDoublets.rds")
```

```{r Annotations}
scrna_YAP = readRDS("/hpc/pmc_kool/fvalzano/Rstudio_Test1/Ependymoma_organoids/Whole/seurat_objs/Post_integration/scrna_YAP_noDoublets.rds")
#condition
Idents(scrna_YAP) = "orig.ident"
scrna_YAP = RenameIdents(scrna_YAP, c("LX360" = "YAP",
                                      "LX362" = "CTRL",
                                      "LX687_YAP" = "YAP",
                                      "LX687_Ctrl" = "CTRL"))
scrna_YAP$condition = scrna_YAP@active.ident
#Overlay annotation from healthy organoids
scrna_YAP = readRDS("/hpc/pmc_kool/fvalzano/Rstudio_Test1/Ependymoma_organoids/Whole/seurat_objs/Post_integration/scrna_YAP_noDoublets.rds")
scrna_healthy = readRDS("/hpc/pmc_kool/fvalzano/Jupyter/scvelo/Ependymoma/rds files/LX_healthy_final.rds")
#Change name of pca slot from pca_new to pca for FindTransferAnchors()
names(scrna_healthy@reductions) = c("pca", "umap")
#Perform SCT normalization and PCA on reference healthy dataset - FindTransferAnchors() works only on embeddings calculated from the same type of normalization
scrna_healthy = SCTransform(scrna_healthy)
scrna_healthy = RunPCA(scrna_healthy)
#Perform reference Map querying
anchors <- FindTransferAnchors(reference = scrna_healthy, query = scrna_YAP, dims = 1:30, reference.reduction = "pca", normalization.method="SCT")
predictions <- TransferData(anchorset = anchors, refdata = scrna$own.mapping, dims = 1:30)
#Add prediction as metadata slot to the single cell object
scrna_YAP <- AddMetaData(scrna_YAP, metadata = predictions)
Idents(scrna_YAP) = "predicted.id"
#Correct naming
scrna_YAP = RenameIdents(scrna_YAP, c("Late RG" = "Late_RG",
                                      "Neuronal" = "Neuronal",
                                      "RG/Astroglia" = "EPN_ECM"))
scrna_YAP$predicted.id = scrna_YAP@active.ident
#Export
write_rds(scrna_YAP, "/hpc/pmc_kool/fvalzano/Rstudio_Test1/Ependymoma_organoids/Whole/seurat_objs/Post_Annotation/scrna_YAP.rds")
```

